File: Assets/StreamingAssets/action.pl
%% Action Ontology - Defines the possible actions that can be performed in the game
%% This file implements the action hierarchy using Prolog rules

%% Dynamic and incremental tabling declarations for action predicates
%% Ensures efficient querying and updates of the action knowledge base
:- table (action/1, interact/1, attack/1, shoot/1, move/1, moveback/1,
     bumpwall/1, pickup/1, discard/1, shootarrow/1) as (incremental, dynamic).

%% Action Hierarchy
%% Defines all possible actions in the game world:
%% - interact: actions involving object manipulation
%% - attack: offensive actions
%% - move: movement actions
%% - moveback: retreat actions
%% - bumpwall: collision actions
action(X) :- interact(X).
action(X) :- attack(X).
action(X) :- move(X).
action(X) :- moveback(X).
action(X) :- bumpwall(X).

%% Interaction Actions
%% Actions that involve manipulating objects in the environment
interact(X) :- pickup(X).    % Picking up items
interact(X) :- discard(X).   % Dropping items

%% Combat Actions
%% Actions related to combat and offensive capabilities
attack(X) :- shoot(X).       % Generic shooting action
shoot(X) :- shootarrow(X).   % Specific bow and arrow attack

File: Assets/Scripts/Ontology/Action.cs
using Agent;
using UnityEngine;

namespace Ontology
{
    public class Action : MonoBehaviour
    {
        public int Utility { get; set; } = 0;
        public virtual void Act() { }
    }

    public class Interact : Action { }

    public class Attack : Action 
    { 
        private AgentAction _agentAction;

        private void Awake() => _agentAction = GetComponent<AgentAction>();

        public override void Act() => _agentAction.TryShootingArrow(); 
    }

    public class BumpWall : Action 
    { 
        private AgentMove _agentMove;

        private void Awake() => _agentMove = GetComponent<AgentMove>();

        public override void Act() => _agentMove.BumpWall(); 
    }

    public class Move : Action 
    { 
        private AgentMove _agentMove;

        private void Awake() => _agentMove = GetComponent<AgentMove>();

        public override void Act() => _agentMove.MoveCell(); 
    }

    public class MoveBack : Action 
    { 
        private AgentMove _agentMove;

        private void Awake() => _agentMove = GetComponent<AgentMove>();

        public override void Act() => _agentMove.MoveAgent(_agentMove.MoveBack()); 
    }

    public class PickUp : Interact 
    { 
        private AgentAction _agentAction;

        private void Awake() => _agentAction = GetComponent<AgentAction>();

        public override void Act() => _agentAction.PickUpGold(); 
    }

    public class Discard : Interact 
    { 
        private AgentAction _agentAction;

        private void Awake() => _agentAction = GetComponent<AgentAction>();

        public override void Act() => _agentAction.Discard(); 
    }

    public class Shoot : Attack { }

    public class ShootArrow : Shoot { }
}